src/kakarot/state.cairo

```diff
@@ -435,17 +439,41 @@ namespace Internals {
     //      Kakarot is not authorized for accounts that are created and SELDESTRUCT in the same transaction
     // @param transfers_len The length of the transfers array.
     // @param transfers The array of Transfer.
-    func _transfer_eth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-        token_address: felt, transfers_len: felt, transfers: model.Transfer*
-    ) {
+    func _transfer_eth{
+        syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, accounts: DictAccess*
+    }(token_address: felt, transfers_len: felt, transfers: model.Transfer*) {
         if (transfers_len == 0) {
             return ();
         }
 
         let transfer = [transfers];
-        IERC20.transferFrom(
-            token_address, transfer.sender.starknet, transfer.recipient.starknet, transfer.amount
-        );
+
+        let (kakarot_address) = get_contract_address();
+
+        let (pointer) = dict_read{dict_ptr=accounts}(key=transfer.recipient.starknet);
+        let recipient = cast(pointer, model.Account*);
+        let is_recipient_registered = Account.is_registered(recipient.address);
+        let use_kakarot_for_recipient = recipient.selfdestruct * (1 - is_recipient_registered);
+        let recipient_address = use_kakarot_for_recipient * kakarot_address + (
+            1 - use_kakarot_for_recipient
+        ) * transfer.recipient.starknet;
+
+        let (pointer) = dict_read{dict_ptr=accounts}(key=transfer.sender.starknet);
+        let sender = cast(pointer, model.Account*);
+        let is_sender_registered = Account.is_registered(sender.address);
+        let use_kakarot_for_sender = sender.selfdestruct * (1 - is_sender_registered);
+        let sender_address = use_kakarot_for_sender * kakarot_address + (
+            1 - use_kakarot_for_sender
+        ) * transfer.sender.starknet;
+
+        // The default ERC20.transferFrom implementation raises even if sender = caller
+        // when there is no prior approval
+        if (sender_address == kakarot_address) {
+            IERC20.transfer(token_address, recipient_address, transfer.amount);
```

Comment:
should we add an invariant check here? something like:
transfer.amount < sender.balance (though this doesn't exist in the current model)

since this performs a direct transfer from Kakarot to some address, this is one of the attack vectors of our architecture:
deploy a contract with tx.value > 0 -> schedule for self destruct in same tx -> contract performs a lot of transfers -> in commit, kakarot will perform the transfers in its place

in the case of some weird wrap around, or failed check when creating a transfer, we might have a delicate situation

that being said, i know that we check sufficient balance before adding to the transfer dict

Reply:
I agree it's not the best. Different solutions I see now are:
- deploying another "escrow" contract instead just for this purpose
- deploying the SELFDESTRUCTed account in any case, but without writing bytecode nor nonce

AFAIU SELFDESTRUCT, it seems like deploying the account and leaving it empty (no code nor nonce) will work for later user

Reply:
> I agree it's not the best. Different solutions I see now are:
> 
> * deploying another "escrow" contract instead just for this purpose
> * deploying the SELFDESTRUCTed account in any case, but without writing bytecode nor nonce
> 
> AFAIU SELFDESTRUCT, it seems like deploying the account and leaving it empty (no code nor nonce) will work for later user
> 
> The benefit or deploying the selfdestructed Starknet account is that it's a one line change and everything else can be kept the same

But it introduces a weird state of 'ghost accounts' where the account is deployed
-> we need to be careful not to try and re-trigger a deploy syscall on the same address in the case of a later deployment, 
-> the account will be registered and "void", so we need to make sure we never use is_registered as a condition for "existing" but rather, hasCodeOrNonce, as in Kakarot SSJ, we extensively (maybe wrongly?) use the is_registered as a condition to be void.

Reply:
`has code or nonce` is the only valid check, and the EF tests eventually the only important things.

as a matter of fact, in this codebase there was only one place where I used `is_registered`, in the beginning of a tx

```
        // Check collision
        let is_registered = Account.is_registered(address.evm);
        let is_collision = is_registered * is_deploy_tx;
```

but I've just replaced with

```
        // Check collision
        let account = Account.fetch_or_create(address);
        let code_or_nonce = Account.has_code_or_nonce(account);
        let is_collision = code_or_nonce * is_deploy_tx;
```